### PR TITLE
[#GEA] Building and floor generalization

### PR DESCRIPTION
**Pull request by lemara98**
Dynamic content has been done. Only equipment, medicine and beds inside rooms are remaining.

-- This code surely needs to be refactored because of DDD

Tip: Try to run the application and play with creating buildings, rooms, and updating the same ones in order to understand the new content better.

### CODE REVIEW

File: GraphicEditor/View/Windows/AddRoom.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        private RoomTypeController roomTypesController = new RoomTypeController();
+        private RoomController roomController = new RoomController();
+
+        public AddRoom(Floor floor1, Border space, FloorUserControl view1)
+        {
+            InitializeComponent();
+            this.DataContext = this;
+            roomSpace = space;
+            floor = floor1;
+            view = view1;
```
Comment (milicasiriski) [inline GraphicEditor/View/Windows/AddRoom.xaml.cs:36]: this.floor = floor and this.view = view and then you don't need 1 in param names, or just name the fields _floor and _view (it's the c# convention)

File: GraphicEditor/ViewModel/MedicineSearchViewModel.cs [hunk 2]
```diff
@@ -89,20 +90,35 @@ private void SearchMedicine(string medicineName)
 
         private void FindMedicine()
         {
+            //Medicine localMedicine = FoundMedicines[SelectedMedicineIndex];
```
Comment (milicasiriski) [inline GraphicEditor/ViewModel/MedicineSearchViewModel.cs:93]: delete if unused

File: project-generated-code-backend/Backend/Repository/Generic/IBuildingRepository.cs [hunk 3]
```diff
@@ -6,5 +6,6 @@ namespace HealthClinicBackend.Backend.Repository.Generic
     public interface IBuildingRepository : IGenericRepository<Building>
     {
         List<Building> GetByName(string name);
+        Building GetBySerialNumber(string serialNumber);
```
Comment (milicasiriski) [inline project-generated-code-backend/Backend/Repository/Generic/IBuildingRepository.cs:9]: This is the GetById serial method in generic repo

File: GraphicEditor/MainWindowViewModel.cs [hunk 4]
```diff
@@ -10,19 +11,19 @@ public class MainWindowViewModel : BindableBase
...
-        public MainWindowViewModel()
+        public MainWindowViewModel(MainWindow _mainWindow)
         {
             NavCommand = new MyICommand<string>(OnNav);
-            
             LoginPage = new LoginUserControl(this);
             HospitalMap = new HospitalMapUserControl(this);
-            CardiologyBuilding = new CardiologyBuildingUserControl(this);
+
+            mainWindow = _mainWindow;
```
Comment (milicasiriski) [inline GraphicEditor/MainWindowViewModel.cs:26]: I don't think you use this anywhere 😄

File: GraphicEditor/MainWindowViewModel.cs [hunk 5]
```diff
@@ -10,19 +11,19 @@ public class MainWindowViewModel : BindableBase
         public MyICommand<string> NavCommand { get; private set; }
         /* TODO add this without causing errors
         private MapContentUserControlViewModel _mapContent = new MapContentUserControlViewModel();*/
```
Comment (Ana00000) [inline GraphicEditor/MainWindowViewModel.cs:13]: Remove this if not needed anymore.

File: GraphicEditor/Resources/Styles/ButtonStyles.xaml [hunk 6]
```diff
@@ -326,14 +412,74 @@
...
+                                  Style="{StaticResource DoorBottomPath}"
+                                  Panel.ZIndex="1">
+                                <Path.RenderTransform>
+                                    <RotateTransform Angle="-90" CenterX="15" CenterY="15"/>
+                                </Path.RenderTransform>
+                            </Path>
+                            <Path HorizontalAlignment="Center"
+                                  VerticalAlignment="Bottom"
+                                  Visibility="{Binding BottomDoor}"
+                                  Style="{StaticResource DoorBottomPath}"
```
Comment (Ana00000) [inline GraphicEditor/Resources/Styles/ButtonStyles.xaml:453]: Excellent work with doors. :))

File: GraphicEditor/View/UserControls/CardiologyFirstFloorMapUserControl.xaml [hunk 7]
```diff
@@ -1,291 +0,0 @@
...
-                   Grid.Row="7" Margin="25,0"/>
-            <Image Style="{StaticResource FlippedChairImage}"
-                   Grid.Column="8"
-                   Grid.ColumnSpan="2"
-                   Grid.Row="7" Margin="25,0"/>
-            <Image Style="{StaticResource FlippedChairImage}"
-                   Grid.Column="8"
-                   Grid.ColumnSpan="2"
-                   Grid.Row="5" Margin="25,0"/>
-            <Image Style="{StaticResource ChairImage}"
```
Comment (Ana00000) [inline GraphicEditor/View/UserControls/CardiologyFirstFloorMapUserControl.xaml:272]: Hopefully, we will implement chairs in the future.

**Review by milicasiriski — APPROVED** [review]

File: GraphicEditor/View/UserControls/FloorUserControl.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+        public FloorUserControl(MainWindowViewModel mapParent, BuildingUserControlViewModel buildingParent, Floor floor)
+        {
+            InitializeComponent();
+            Viewmodel = new FloorUserControlViewModel(mapParent, buildingParent, this.Grid, floor);
+            this.DataContext = Viewmodel;
+        }
+
+        private void Grid_MouseLeftButtonUp(object sender, MouseButtonEventArgs e)
+        {
```
Comment (Ana00000) [inline GraphicEditor/View/UserControls/FloorUserControl.xaml.cs:30]: Perhaps you can split this function into a few.

File: GraphicEditor/View/UserControls/FloorUserControl.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+            double accumulatedWidth = 0.0;
+
+            // calc row mouse was over
+            foreach (var rowDefinition in Grid.RowDefinitions)
+            {
+                accumulatedHeight += rowDefinition.ActualHeight;
+                if (accumulatedHeight >= mouseUpPos.Y)
+                    break;
+                row++;
+            }
```
Comment (Ana00000) [inline GraphicEditor/View/UserControls/FloorUserControl.xaml.cs:53]: For example function 1

File: GraphicEditor/View/UserControls/FloorUserControl.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+            // calc col mouse was over
+            foreach (var columnDefinition in Grid.ColumnDefinitions)
+            {
+                accumulatedWidth += columnDefinition.ActualWidth;
+                if (accumulatedWidth >= mouseUpPos.X)
+                    break;
+                col++;
+            }
+
```
Comment (Ana00000) [inline GraphicEditor/View/UserControls/FloorUserControl.xaml.cs:63]: For example function 2

File: GraphicEditor/View/UserControls/FloorUserControl.xaml.cs [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            Grid.SetColumnSpan(selectedGridCells, span.Item1);
+            Grid.SetRowSpan(selectedGridCells, span.Item2);
+
+            SelectedCellsColumn = firstGridCell.Item1;
+            SelectedCellsRow = firstGridCell.Item2;
+            SelectedCellsColumnSpan = span.Item1;
+            SelectedCellsRowSpan = span.Item2;
+
+            selectedGridCells.Visibility = Visibility.Visible;
```
Comment (Ana00000) [inline GraphicEditor/View/UserControls/FloorUserControl.xaml.cs:82]: For example function 3

File: GraphicEditor/View/Windows/AddRoom.xaml.cs [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            Grid.SetZIndex(newRoomButton, 2);
+            view.Grid.Children.Add(newRoomButton);
+            view.Grid.UpdateLayout();
+
+            roomController.NewRoom(newRoom);
+
+            roomSpace.Visibility = Visibility.Collapsed;
+            this.Close();
+        }
+
```
Comment (Ana00000) [inline GraphicEditor/View/Windows/AddRoom.xaml.cs:121]: Function is too big, split it.

File: GraphicEditor/ViewModel/AddBuildingViewModel.cs [hunk 13]
```diff
@@ -57,7 +65,38 @@ private void AddBuilding(object obj)
...
         {
             Button.Style = (Style)myResourceDictionary[buildingButtonStyle];
-            Button.Background = new SolidColorBrush(MyButtonColor);
+            Button.Background = new SolidColorBrush(MyButtonColor); 
+            Button.Content = NameText;
+
+            string content = (string)Button.Tag;
+            string[] split = content.Split(" ");
+
+            // [Lemara98] This needs to be corrected because I didn't find appropriate controller for buildings !!!
```
Comment (Ana00000) [inline GraphicEditor/ViewModel/AddBuildingViewModel.cs:74]: There exists controller for adding new buildings.

Comment (lemara98) [reply GraphicEditor/ViewModel/AddBuildingViewModel.cs:74]: I made some changes to controller (added some that were missing) and services (added some more complexed functionalities)

**Pull request closed without merging**